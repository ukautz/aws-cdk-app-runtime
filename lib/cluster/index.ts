export * from './cluster';
export * from './router';
export { specKeys as clusterSpecKeys, Specs as ClusterSpecs } from './specs';
