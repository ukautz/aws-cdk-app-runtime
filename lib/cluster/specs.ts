import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecs from '@aws-cdk/aws-ecs';
import * as route53 from '@aws-cdk/aws-route53';
import * as servicediscovery from '@aws-cdk/aws-servicediscovery';
import { keys as compiledKeysOf } from 'ts-transformer-keys';
import { SpecUtil } from '../util/spec';
import { RouterSpecs } from './router';

/**
 * Specs (State) that the runtime cluster exports, that will be required by other
 * runtime components within the runtime cluster, like runtime services.
 */
export interface Specs extends RouterSpecs {
  /**
   * Name of the CloudMap namespace, i.e. the cluster internal hostname
   */
  privateDomain: string;

  /**
   * The full public domain names, like 'apps.domain.tld', of which sub-domains like '<my-app>.apps.domain.tld>', will be routing traffic to service <my-app>
   */
  publicDomain: string;

  /**
   * ARN of the hosted zone that holds the apex and wildcard records pointing to the load balancer
   */
  zoneArn: string;

  /**
   * ID of the hosted zone that holds the apex and wildcard records pointing to the load balancer
   */
  zoneId: string;

  /**
   * Name of the hosted zone that holds the apex and wildcard records pointing to the load balancer
   */
  zoneName: string;

  /**
   * Comma separated list of availability zones this cluster is using
   */
  availabilityZones: string;

  /**
   * ARN of the ECS cluster
   */
  clusterArn: string;

  /**
   * Name of the ECS cluster
   */
  clusterName: string;

  /**
   * ARN of the cloudmap namespace for internal service name resolution
   */
  namespaceArn: string;

  /**
   * ID of the cloudmap namespace for internal service name resolution
   */
  namespaceId: string;

  /**
   * Name of the cloudmap namespace for internal service name resolution
   */
  namespaceName: string;

  /**
   * ID of the VPC the ECS cluster runs in
   */
  vpcId: string;
}

/**
 * Names of keys in the Runtime Cluster Specification
 */
export const specKeys = compiledKeysOf<Specs>().sort();

export class Specs implements Specs {
  static make = (generator?: (prop: string) => string): Specs => SpecUtil.make<Specs>(specKeys, generator);
  static fromContext = (scope: cdk.Construct, prefix?: string): Specs => SpecUtil.fromContext(specKeys, scope, prefix);
  static fromSsm = (scope: cdk.Construct, prefix?: string): Specs => SpecUtil.fromSsm<Specs>(specKeys, scope, prefix);
  static toSsm = (scope: cdk.Construct, prefix: string, specs: Specs, secure?: boolean): void =>
    SpecUtil.toSsm<Specs>(scope, prefix, specs, secure);

  /**
   * Loads VPC defined in cluster specs
   */
  static lookupVpc = (scope: cdk.Construct, id: string, specs: Specs): ec2.IVpc =>
    ec2.Vpc.fromLookup(scope, id, {
      vpcId: specs.vpcId,
    });

  /**
   * Loads ECS cluster defined in cluster specs
   */
  static lookupCluster = (scope: cdk.Construct, id: string, specs: Specs, vpc?: ec2.IVpc): ecs.ICluster =>
    ecs.Cluster.fromClusterAttributes(scope, id, {
      clusterName: specs.clusterName,
      securityGroups: [],
      vpc: vpc ?? Specs.lookupVpc(scope, `${id}Vpc`, specs),
    });

  /**
   * Loads cluster cloudmap / service discovery namespace (for private domain)
   */
  static lookupPrivateNamespace = (
    scope: cdk.Construct,
    id: string,
    specs: Specs
  ): servicediscovery.IPrivateDnsNamespace =>
    servicediscovery.PrivateDnsNamespace.fromPrivateDnsNamespaceAttributes(scope, id, specs);
}
