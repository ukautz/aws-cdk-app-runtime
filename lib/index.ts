export * from './util/index';
export * from './cluster/index';
export * from './service/index';
