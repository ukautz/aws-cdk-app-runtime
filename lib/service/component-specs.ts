export interface ComponentSpecs {
  /**
   * Name of the cluster that this component runs in
   */
  cluster: string;

  /**
   * The ARN of the IAM role / principal that the runtime provides to the running Service
   */
  iamPrincipal: string;

  /**
   * Runtime resource and property requirements of this component
   */
  resources: string;

  /**
   * Security groups that the executed Service has, separated by ","
   */
  securityGroupIds: string;
}
