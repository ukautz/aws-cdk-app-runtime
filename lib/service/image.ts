import { aws_ecr as ecr, aws_ecs as ecs } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as fs from 'fs';
import * as path from 'path';

/**
 *
 * @param scope
 * @param imageName can be `<dockerhub image>` or `ecr:<arn>` or `path:</to/folder/with/dockerfile>`
 * @param imageVersion for docker hub image: added version tag, for ecr: added version arn, for local path: value of BUILD_TAG build parameter
 */
export const loadContainerImage = (
  scope: Construct,
  imageName: string,
  imageVersion?: string,
  buildArgs?: Record<string, string>
): { image: ecs.ContainerImage; repository: ecr.IRepository | undefined } => {
  let repository: ecr.IRepository | undefined;
  let image: ecs.ContainerImage;

  // load from file
  if (imageName.startsWith('path:')) {
    let directory = imageName.substr('path:'.length);
    let opt: Partial<ecs.AssetImageProps> = {};

    // can point to directory containing Dockerfile OR pointing to specific CustomDockerfile
    if (fs.statSync(directory).isFile()) {
      const file = path.basename(directory);
      directory = path.dirname(directory);
      opt = { file };
    }

    image = ecs.ContainerImage.fromAsset(directory, {
      ...opt,
      buildArgs: {
        ...(buildArgs ?? {}),
        BUILD_TAG: imageVersion ?? 'latest',
      },
    });

    // load from ECR repository
  } else if (imageName.startsWith('ecr:')) {
    const name = imageName.substr('ecr:'.length);
    const repoName = parseEcrImageName(name);
    repository = ecr.Repository.fromRepositoryName(scope, 'Repository', repoName);
    image = ecs.ContainerImage.fromEcrRepository(repository, imageVersion);

    // load from DockerHub or another online registry
  } else {
    image = ecs.ContainerImage.fromRegistry(imageName + ':' + imageVersion);
  }
  return { image, repository };
};

/**
 * Parses ECR image URL into ECR image repo ARN and image tag
 *
 * @param image - the image ECR URL, like `123123123.dkr.ecr.xx-region-1.amazonaws.com/repo-name:image-version-tag`
 * @returns the image repository name (like `arn:aws:ecr:xx-region-1:123123123:repository/repo-name`) and the image tag (like `v1.2.3`)
 */
function parseEcrImageName(image: string): string {
  const match = image.match(/^([0-9]+)\.dkr\.ecr\.([^.]+)\.amazonaws\.com\/([^:]+)$/);
  if (!match) {
    throw new Error(`invalid ECR image URL "${image}" (hint: must not contain tag)`);
  }

  //arn:aws:ecr:xx-region-1:123123123:repository/repo-name
  return `arn:aws:ecr:${match[2]}:${match[1]}:repository/${match[3]}`;
}
