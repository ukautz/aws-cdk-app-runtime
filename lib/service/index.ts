export * from './component';
export * from './component-specs';
export * from './image';
export * from './scaler';
export * from './service';
export * from './service-specs';
export * from './task';
export * from './task-specs';
