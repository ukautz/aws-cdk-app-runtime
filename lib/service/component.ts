import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as ecs from '@aws-cdk/aws-ecs';
import * as iam from '@aws-cdk/aws-iam';
import * as ssm from '@aws-cdk/aws-ssm';
import { Cluster, ClusterSpecs } from '../cluster';
import { Resources, ResourcesProps } from '../util';
import { loadContainerImage } from './image';

export interface ComponentProps {
  /**
   * The name of the service (unique in the cluster namespace and stage).
   *
   * Will be used to expose service externally and internally (see `exposed`)
   */
  name: string;

  /**
   * Can be `<dockerhub image>` or `ecr:<arn>` or `path:</to/folder/with/dockerfile>`
   */
  image: string | ecs.ContainerImage;

  /**
   * For dockerhub: added version tag
   * For ECR: added as version suffix to image ARN
   * For Local path: Added BUILD_TAG build argument
   */
  imageVersion?: string;

  /**
   * The docker build arguments to pass to the build process (if `image` property references a docker file containing directory)
   */
  dockerBuildArgs?: Record<string, string>;

  /**
   * The resources the service should get
   */
  resources?: ResourcesProps;

  /**
   * Optional environment variables in the form `{"ENV_NAME": "env value"}`, to be passed to container
   */
  environment?: Record<string, string>;
  /**
   * Optional secret environment variables in the form `{"ENV_NAME": "arn:ssm:..:key"}`, to be passed to container
   */
  secrets?: Record<string, string>;

  /**
   * Optional logging setup of container output
   */
  logging?: false | string | ecs.LogDriver;

  /**
   * Cluster specs can be either provided directly or as a prefix string to load from SSM
   */
  cluster: ClusterSpecs | Cluster | string | ((scope: cdk.Construct) => { cluster: ecs.ICluster; specs: ClusterSpecs });
}

/**
 * Super class for any Component that runs in a Cluster. Shared properties of Services and Tasks.
 */
export abstract class Component<ComponentT extends ComponentProps> extends cdk.Construct implements iam.IGrantable {
  /**
   * The name of the Component
   */
  public readonly name: string;

  /**
   * The Cluster this Component runs in
   */
  public readonly cluster: ecs.ICluster;

  /**
   * The specifications of the Cluster that this Component runs in
   */
  public readonly clusterSpecs: ClusterSpecs;

  /**
   * The auth principal for execution the running container of the Service will provide.
   * Should be granted access to databases and such.
   */
  public readonly taskRole: iam.Role;

  /**
   * The (docker) container image that implements this Component
   */
  public readonly image: ecs.ContainerImage;

  /**
   * Auth principle that executes (starts) the Component (containers).
   * MUST have access to repository that contains image.
   */
  public readonly taskExecutionRole: iam.Role;

  /**
   * The resources required by this Component
   */
  public readonly resources: Resources;

  public readonly taskDefinition: ecs.FargateTaskDefinition;

  constructor(scope: cdk.Construct, id: string, props: ComponentT) {
    super(scope, id);

    // initialize cluster specs
    this.name = props.name;

    if (typeof props.cluster === 'string') {
      this.clusterSpecs = ClusterSpecs.fromSsm(this, props.cluster);
    } else if (typeof props.cluster === 'function') {
      const { cluster, specs } = props.cluster(this);
      this.clusterSpecs = specs;
      this.cluster = cluster;
    } else if (props.cluster instanceof Cluster) {
      this.clusterSpecs = props.cluster.specs;
      this.cluster = props.cluster.cluster;
    } else {
      this.clusterSpecs = props.cluster;
    }
    if (!this.cluster) {
      this.cluster = specClusterLoader(this, this.clusterSpecs);
    }
    this.resources = new Resources(props.resources);
    this.taskRole = this.initTaskRole(props);
    this.taskExecutionRole = this.initTaskExecutionRole(props);
    this.image = this.initImage(props);
    this.taskDefinition = this.initTaskDefinition(props);

    function specClusterLoader(scope: cdk.Construct, specs: ClusterSpecs): ecs.ICluster {
      return ClusterSpecs.lookupCluster(scope, 'Cluster', specs);
    }
  }

  public get grantPrincipal(): iam.IPrincipal {
    return this.taskRole;
  }

  private initImage(props: ComponentProps): ecs.ContainerImage {
    if (typeof props.image !== 'string') {
      return props.image;
    }
    const { image, repository } = loadContainerImage(this, props.image, props.imageVersion, props.dockerBuildArgs);
    repository?.grantPull(this.taskExecutionRole);
    return image;
  }

  /**
   * See https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html
   *
   * @param props
   * @returns
   */
  private initTaskRole(props: ComponentT): iam.Role {
    return new iam.Role(this, 'TaskRole', {
      description: `role ${props.name} is provided to assume from within`,
      assumedBy: new iam.ServicePrincipal('ecs-tasks.amazonaws.com', {
        /* conditions: {
          StringEquals: {
            'ecs:cluster': this.cluster.clusterArn,
          },
        }, */
      }),
    });
  }

  /**
   * See https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_execution_IAM_role.html
   * @param props
   * @returns
   */
  private initTaskExecutionRole(props: ComponentT): iam.Role {
    return new iam.Role(this, 'TaskExecutionRole', {
      description: `role assumed by ECS to start ${props.name}`,
      managedPolicies: [iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AmazonECSTaskExecutionRolePolicy')],
      assumedBy: new iam.ServicePrincipal('ecs-tasks.amazonaws.com', {
        /* conditions: {
          StringEquals: {
            'ecs:cluster': this.cluster.clusterArn,
          },
        }, */
      }),
    });
  }

  private initTaskDefinition(props: ComponentT): ecs.FargateTaskDefinition {
    return new ecs.FargateTaskDefinition(this, 'TaskDefinition', {
      cpu: this.resources.cpu,
      memoryLimitMiB: this.resources.memory,
      taskRole: this.taskRole,
      executionRole: this.taskExecutionRole,
    });
  }
}

export const secretsFromProps = (scope: cdk.Construct, secrets?: Record<string, string>): Record<string, ecs.Secret> =>
  Object.fromEntries(
    Object.entries(secrets ?? {}).map(([key, value]) => {
      return [
        key,
        ecs.Secret.fromSsmParameter(
          ssm.StringParameter.fromStringParameterName(scope, `ServiceContainerEnv${key}`, value)
        ),
      ];
    })
  );
