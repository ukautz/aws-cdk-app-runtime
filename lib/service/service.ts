import * as cdk from 'aws-cdk-lib';
import { aws_ec2 as ec2, aws_ecs as ecs, aws_servicediscovery as servicediscovery } from 'aws-cdk-lib';
import { Construct, IConstruct } from 'constructs';
import { ClusterSpecs } from '../cluster';
import { ecsLogDriver } from '../util/logging';
import { Component, ComponentProps, secretsFromProps } from './component';
import { Scaler } from './scaler';
import { ServiceSpecs } from './service-specs';

export interface ServiceProps extends ComponentProps {
  /**
   * Whether this component is exposed.
   * As a service that would be as an load balanced network endpoint at `https://<service>.public.domain`, in any case it would be available from within the cluster as `<service>.private.domain`.
   * As a task:
   * - has a Lambda interface, that allows to start, stop and get status of the task
   * - public means provides a REST API at `https://<service>.public.domain`,
   */
  public: boolean;

  /**
   * The port the container is exposing (when public==true), to which HTTP traffic is to be routed
   *
   * Default: 80
   */
  containerPort?: number;
}

/**
 * Aspect that can be applied to services forcing WEIGHTED routing policy.
 *
 * This is a HACK-AROUND since FargateService API does not allow to set routing policy, which
 * then defaults to the default, which is MULTIVALUE
 *
 * @see https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ecs.FargateService.html#cloudmapoptions - misses option to set routing policy
 */
class EnforceWeightedRoutingForService implements cdk.IAspect {
  public visit(node: IConstruct): void {
    if (!(node instanceof servicediscovery.CfnService)) {
      return;
    }
    node.dnsConfig = {
      ...(node.dnsConfig as servicediscovery.CfnService.DnsConfigProperty),
      routingPolicy: servicediscovery.RoutingPolicy.WEIGHTED,
    };
  }
}

/**
 * A runtime service runs continuously and never stops. Think of a web service, mail service etc.
 */
export class Service extends Component<ServiceProps> implements ec2.IConnectable {
  public readonly service: ecs.FargateService;

  constructor(scope: Construct, id: string, props: ServiceProps) {
    super(scope, id, props);

    const mode = props.resources?.scaling?.mode ?? 'fixed';
    if (!['fixed', 'scaling'].includes(mode)) {
      throw new Error(`Service requires resource scaling mode "fixed" or "scaling", unsupported: "${mode}"`);
    }

    this.service = this.initService(props);
    this.setupAutoScaling();
    this.setupRoutingPermissions(props);
  }

  public get specs(): ServiceSpecs {
    return {
      cluster: this.cluster.clusterName,
      iamPrincipal: this.taskRole.roleArn,
      privateHostname: `${this.name}.${this.clusterSpecs.privateDomain}`,
      publicHostname: `${this.name}.${this.clusterSpecs.publicDomain}`,
      resources: this.resources.toString(),
      securityGroupIds: this.service.connections.securityGroups
        .map((sg) => sg.securityGroupId)
        .sort()
        .join(','),
      service: this.name,
    };
  }

  public get connections(): ec2.Connections {
    return this.service.connections;
  }

  /**
   * Create service in ECS cluster
   */
  private initService(props: ServiceProps): ecs.FargateService {
    // enable traffic for public services only
    const traefikLabels = props.public
      ? {
          'traefik.enable': 'true',
          [`traefik.http.routers.${props.name}.rule`]: `Host(\`${props.name}.${this.clusterSpecs.publicDomain}\`)`,
        }
      : {
          'traefik.enable': 'false',
        };

    // attach docker container to task
    const container = this.taskDefinition.addContainer('ServiceContainer', {
      image: this.image,
      logging: ecsLogDriver(props.logging ?? `/service/${props.name}`),
      environment: props.environment,
      secrets: secretsFromProps(this, props.secrets),
      dockerLabels: {
        'runtime.resources': this.resources.toString(),
        ...traefikLabels,
      },
    });

    // port mapping exposes the container
    const port = props.containerPort ?? 80;
    container.addPortMappings({
      containerPort: port,
      hostPort: port,
      protocol: ecs.Protocol.TCP,
    });

    // for private address, require (cloudmap / servicediscovery) namespace
    const cloudMapNamespace =
      this.cluster.defaultCloudMapNamespace ??
      ClusterSpecs.lookupPrivateNamespace(this, 'PrivateNamespace', this.clusterSpecs);

    const service = new ecs.FargateService(this, 'Service', {
      cluster: this.cluster,
      taskDefinition: this.taskDefinition,
      cloudMapOptions: {
        cloudMapNamespace,
        name: props.name,
      },
      ...(this.resources.scaling.mode === 'fixed' ? { desiredCount: this.resources.scaling.amount } : {}),
    });

    // weighted routing gives us a poor mans load balancer - makes more sense than
    // every client having to implement that itself by MULTIVALUE ip address set
    // @see
    cdk.Aspects.of(service).add(new EnforceWeightedRoutingForService());

    return service;
  }

  /**
   * Grant traffic routing from load balancer
   */
  private setupRoutingPermissions(props: ServiceProps): void {
    // only grant routing permissions to router if public routing is enabled
    if (!props.public) {
      return;
    }

    // load connection to router service, to grant it's permission to route traffic to this service
    const routerSecurityGroups = this.clusterSpecs.routerSecurityGroupIds
      .split(',')
      .map((id, idx) => ec2.SecurityGroup.fromSecurityGroupId(this, `SecurityGroup${idx + 1}`, id));
    const routerConnections = new ec2.Connections({
      securityGroups: routerSecurityGroups,
    });

    // traffic from router may flow to service
    this.service.connections.allowFrom(
      routerConnections,
      new ec2.Port({
        protocol: ec2.Protocol.TCP,
        fromPort: 80,
        toPort: 80,
        stringRepresentation: 'HTTP',
      })
    );
  }

  /**
   * Configures services scaling configuration
   */
  private setupAutoScaling(): void {
    if (this.resources.scaling.mode !== 'scaling') return;
    new Scaler(this, 'Scaler', {
      scaling: this.resources.scaling,
      service: this.service,
    });
  }
}
