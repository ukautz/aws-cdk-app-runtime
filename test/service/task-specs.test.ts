import '@aws-cdk/assert/jest';
import { TaskSpecs, taskSpecKeys } from '../../lib/service';
import { assertSpecs } from '../util/specs.suite';

describe('Task Specs', () => {
  test('Can list keys', () => {
    expect(taskSpecKeys).toEqual(['arn', 'cluster', 'iamPrincipal', 'resources', 'securityGroupIds', 'task'].sort());
  });
  assertSpecs<TaskSpecs>({
    name: 'Task',
    keys: taskSpecKeys,
    make: TaskSpecs.make,
    fromContext: TaskSpecs.fromContext,
    fromSsm: TaskSpecs.fromSsm,
    toSsm: TaskSpecs.toSsm,
  });
});
