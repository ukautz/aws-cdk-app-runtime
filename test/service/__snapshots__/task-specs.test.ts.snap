// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task Specs Task Store Encrypted to Ssm Matches Snapshot 1`] = `
Object {
  "Resources": Object {
    "ArnSpecE033A0CE": Object {
      "Properties": Object {
        "Name": "/prefixed/arn",
        "Type": "SecureString",
        "Value": "arn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ClusterSpecD8C9F318": Object {
      "Properties": Object {
        "Name": "/prefixed/cluster",
        "Type": "SecureString",
        "Value": "cluster",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "IamPrincipalSpec6C2077B5": Object {
      "Properties": Object {
        "Name": "/prefixed/iamPrincipal",
        "Type": "SecureString",
        "Value": "iamPrincipal",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ResourcesSpec1B6B0F2A": Object {
      "Properties": Object {
        "Name": "/prefixed/resources",
        "Type": "SecureString",
        "Value": "resources",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SecurityGroupIdsSpec5B6C3A31": Object {
      "Properties": Object {
        "Name": "/prefixed/securityGroupIds",
        "Type": "SecureString",
        "Value": "securityGroupIds",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "TaskSpecB01B854A": Object {
      "Properties": Object {
        "Name": "/prefixed/task",
        "Type": "SecureString",
        "Value": "task",
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;

exports[`Task Specs Task Store to Ssm Matches Snapshot 1`] = `
Object {
  "Resources": Object {
    "ArnSpecE033A0CE": Object {
      "Properties": Object {
        "Name": "/prefixed/arn",
        "Type": "String",
        "Value": "arn",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ClusterSpecD8C9F318": Object {
      "Properties": Object {
        "Name": "/prefixed/cluster",
        "Type": "String",
        "Value": "cluster",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "IamPrincipalSpec6C2077B5": Object {
      "Properties": Object {
        "Name": "/prefixed/iamPrincipal",
        "Type": "String",
        "Value": "iamPrincipal",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "ResourcesSpec1B6B0F2A": Object {
      "Properties": Object {
        "Name": "/prefixed/resources",
        "Type": "String",
        "Value": "resources",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "SecurityGroupIdsSpec5B6C3A31": Object {
      "Properties": Object {
        "Name": "/prefixed/securityGroupIds",
        "Type": "String",
        "Value": "securityGroupIds",
      },
      "Type": "AWS::SSM::Parameter",
    },
    "TaskSpecB01B854A": Object {
      "Properties": Object {
        "Name": "/prefixed/task",
        "Type": "String",
        "Value": "task",
      },
      "Type": "AWS::SSM::Parameter",
    },
  },
}
`;
